### PR TITLE
Use isArray instead of array supertype checks

### PR DESCRIPTION
**Pull request by herndlm**
another ~~small~~ (got bigger over time..) refactor. I'll stop now.. :)

UPDATE: added another commit that does the same for isString

### CODE REVIEW

**Comment by herndlm** [comment]
I was not expecting that one additional failure. weird. needs some debugging..

**Comment by staabm** [comment]
one more here
https://github.com/phpstan/phpstan-src/blob/ed6548cb2db72dad716d9920805dc5924a1ea932/src/Analyser/TypeSpecifier.php#L529

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -1474,11 +1474,13 @@ private function resolveType(Expr $node): Type
...
-			} elseif ($stringType->isSuperTypeOf($varType)->yes()) {
+			} elseif ($varType->isString()->yes()) {
 				if ($varType->isLiteralString()->yes()) {
-					return new IntersectionType([$stringType, new AccessoryLiteralStringType()]);
+					return new IntersectionType([new StringType(), new AccessoryLiteralStringType()]);
 				}
-				return $stringType;
+				return new StringType();
+			} elseif ($neverType->isSuperTypeOf($varType)->yes()) {
+				return $neverType;
```
Comment (herndlm) [inline src/Analyser/MutatingScope.php]: I don't quite like this. The problem is that with NeverType being the bottom type all isSuperTypeOf would always return yes. I think I could solve this cleaner if Type had more methods like isInteger, isFloat and isNull or so, but not sure if it's really worth it.

Comment (herndlm) [reply src/Analyser/MutatingScope.php]: extracted into https://github.com/phpstan/phpstan-src/pull/1641
the change here will go away again

Comment (ondrejmirtes) [reply src/Analyser/MutatingScope.php]: Yeah, I plan to add a lot of methods like these, they'd be really handy.

**Comment by ondrejmirtes** [comment]
Please ping me once you mark this as ready, I'm looking forward to merge this.

**Comment by herndlm** [comment]
I think in order to be able to continue here I'd need https://github.com/phpstan/phpstan-src/pull/1644 to avoid weird behaviour or exceptions

**Comment by herndlm** [comment]
@ondrejmirtes this is ready now. I could create `isInteger` next and do the same if you like that

**Comment by herndlm** [comment]
@ondrejmirtes this is ready now. I could create `isInteger` next and do the same if you like that

**Comment by ondrejmirtes** [comment]
Yeah, you could be the guy who spends the next 6 months painstakingly creating methods on Type interface for all the use cases where we currently use instanceof *Type and TypeUtils methods :) That would be really awesome if it sounds interesting to you :)

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by herndlm** [comment]
> Yeah, you could be the guy who spends the next 6 months painstakingly creating methods on Type interface for all the use cases where we currently use instanceof *Type and TypeUtils methods :) That would be really awesome if it sounds interesting to you :)

you need to work on your sales pitch :D
I don't want to promise much, but I'll definitely continue with some primitive type methods. Feel free to mention any methods you're interested in

**Comment by ondrejmirtes** [comment]
It's not my fault there's "pain" in "painstakingly" 🤣

**Comment by ondrejmirtes** [comment]
Another easy one should be to replace stuff like `instanceof TypeWithClassName`, `instanceof ObjectType`, `TypeUtils::getDirectClassNames()`.

**Comment by herndlm** [comment]
AFAIK the problems with instanceof replacements are then that methods are not guaranteed to exist on the Type interface, right? E.g. it's not safe anymore to call `getClassName()` on the potential union :/

ah and I misread "painstakingly" with "painfully" yesterday I guess. Both are fitting terms I think :D

**Comment by ondrejmirtes** [comment]
This is the problem I'm trying to solve - there needs to be methods for EVERYTHING directly on `Type`. It nicely solves issues when the type is `UnionType` etc.

**Comment by herndlm** [comment]
definitely, yeah. it would also surely help with some constantarray operations I think and then greatly simplify the code that currently needs conditionals to deal with them with the downside of blowing up the Type interface I guess. but for the array/constantarray methods there could be a trait that implements the array-specific methods once for all non-array types or so and then maybe it's not a big deal after all

**Comment by ondrejmirtes** [comment]
What we can do once the transformation is complete is to remove all specific `Template*Type` implementations and have just one `TemplateType` (class!) that supports any bound.

We can also stop inheriting type classes from each other.

And I guess we could also have a general `SubtractableType` that supports any type.

**Pull request merged**